catalog/files/impl/src/main/java/org/projectnessie/catalog/files/adls/AdlsProgrammaticOptions.java

```diff
@@ -15,117 +15,57 @@
 
   @Override
-  OptionalLong writeBlockSize();
-
-  @Override
-  OptionalInt readBlockSize();
-
-  @Override
-  Map<String, String> configurationOptions();
-
-  @Override
-  Map<String, AdlsFileSystemOptions> fileSystems();
-
-  static Builder builder() {
-    return ImmutableAdlsProgrammaticOptions.builder();
-  }
-
-  @SuppressWarnings("unused")
-  interface Builder {
-    @CanIgnoreReturnValue
-    Builder readBlockSize(int readBlockSize);
-
-    @CanIgnoreReturnValue
-    Builder writeBlockSize(long writeBlockSize);
-
-    @CanIgnoreReturnValue
-    Builder putConfigurationOptions(String key, String value);
-
-    @CanIgnoreReturnValue
-    Builder putConfigurationOptions(Map.Entry<String, ? extends String> entry);
-
-    @CanIgnoreReturnValue
-    Builder configurationOptions(Map<String, ? extends String> entries);
-
-    @CanIgnoreReturnValue
-    Builder putAllConfigurationOptions(Map<String, ? extends String> entries);
-
-    @CanIgnoreReturnValue
-    Builder putFileSystems(String key, AdlsFileSystemOptions value);
-
-    @CanIgnoreReturnValue
-    Builder putFileSystems(Map.Entry<String, ? extends AdlsFileSystemOptions> entry);
-
-    @CanIgnoreReturnValue
-    Builder fileSystems(Map<String, ? extends AdlsFileSystemOptions> entries);
-
-    @CanIgnoreReturnValue
-    Builder putAllFileSystems(Map<String, ? extends AdlsFileSystemOptions> entries);
-
-    @CanIgnoreReturnValue
-    Builder defaultOptions(AdlsFileSystemOptions defaultOptions);
-
-    AdlsProgrammaticOptions build();
+  Map<String, AdlsNamedFileSystemOptions> fileSystems();
+
+  static AdlsProgrammaticOptions copyOf(AdlsOptions adlsOptions) {
+    ImmutableAdlsProgrammaticOptions.Builder builder =
+        ImmutableAdlsProgrammaticOptions.builder().from(adlsOptions);
+    // not copied by from() because of different type parameters in method return types
+    builder.defaultOptions(adlsOptions.defaultOptions());
+    builder.fileSystems(adlsOptions.fileSystems());
+    return builder.build();
   }
 
-  @Value.Immutable
-  interface AdlsPerFileSystemOptions extends AdlsFileSystemOptions {
-
-    static Builder builder() {
-      return ImmutableAdlsPerFileSystemOptions.builder();
+  @Value.Check
+  default AdlsProgrammaticOptions normalizeBuckets() {
+    Map<String, AdlsNamedFileSystemOptions> fileSystems = new LinkedHashMap<>();
+    for (String fsKey : fileSystems().keySet()) {
+      AdlsNamedFileSystemOptions options = fileSystems().get(fsKey);
+      if (options.name().isPresent()) {
+        fsKey = options.name().get();
+      } else {
+        options = ImmutableAdlsNamedFileSystemOptions.builder().from(options).name(fsKey).build();
+      }
+      if (fileSystems.put(fsKey, options) != null) {
+        throw new IllegalArgumentException("Duplicate filesystem name: " + fsKey);
+      }
     }
+    if (fileSystems.equals(fileSystems())) {
+      return this;
+    }
+    return ImmutableAdlsProgrammaticOptions.builder().from(this).fileSystems(fileSystems).build();
```

Comment:
```suggestion
    return ImmutableAdlsProgrammaticOptions.builder().from(this).fileSystems(fileSystems)
      .defaultOptions(defaultOptions())
      .build();
```

Reply:
I think this isn't required because in this case, `defaultOptions` will be properly copied. But it doesn't hurt to add it.

Reply:
Hm - but file-system would be copied as well. Confused.

Reply:
Yes but here we want to overwrite the `fileSystems` attribute that was copied from `this` with the normalized `fileSystems` variable.